recipes-kernel/linux/5.4/patches/blktap2.patch

```diff
@@ -391,23 +391,23 @@ PATCHES
 +{
 +	struct blktap *tap;
 +	struct blktap_ring *ring;
++	int r;
 +
 +	tap = blktaps[minor];
 +	if (!tap)
 +		return -ENODEV;
 +
 +	ring = &tap->ring;
 +
-+	for (;;) {
-+		int r;
++	if (ring->vma == NULL && ring->n_pending)
++		printk(KERN_WARNING
++		       "blktap%d ring not mapped and n_pending %d\n",
++		       minor, ring->n_pending);
```

Comment:
Greping through some logs,  I see values from -1 all the way up to 103.  I think this needs some more investigation. :(

-1 points toward an accounting issue somewhere.  blktap_ring_free_request() doesn't check if `ring->pending[request->usr_idx] != NULL`, so that could be it.
In general, this is surprising, but it isn't a problem because they are all read-only.  It's the hashing of read-only .vhds and extraction of kernels.

I'm concerned that a read-write .vhd could have the tapdisk exit while there at still pending writes.  I think, on close, the blk-mq stuff needs to be stopped, tapdisk needs to process everything, and only then tapdisk can detach.

Reply:
I added a check after "if (ring->vma == NULL && ring->n_pending)" to iterate ring->requests and count the number of non-NULL entries.  The result was 0, not matching n_pending.

I switched n_pending to atomic_t and the n_pending count stays accurate.  I haven't looked to see where it may be going wrong.